packages/flutter/lib/src/widgets/restoration.dart

```diff
@@ -0,0 +1,1031 @@
+
+  /// The widget below this widget in the tree.
+  ///
+  /// {@macro flutter.widgets.child}
+  final Widget child;
+
+  /// The [RestorationId] used to identify the child bucket that this widget
+  /// will insert into the tree.
+  ///
+  /// If this is null, no bucket is made available to descendants and state
+  /// restoration for the subtree is essentially turned off.
+  final RestorationId restorationId;
+
+  @override
+  State<RootRestorationScope> createState() => _RootRestorationScopeState();
+}
+
+class _RootRestorationScopeState extends State<RootRestorationScope> {
+  bool _okToRenderBlankContainer;
+  bool _rootBucketValid = false;
+  RestorationBucket _rootBucket;
+  RestorationBucket _ancestorBucket;
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    _ancestorBucket = RestorationScope.of(context);
+    _loadRootBucketIfNecessary();
+    _okToRenderBlankContainer ??= widget.restorationId != null && _needsRootBucketInserted;
+  }
+
+  @override
+  void didUpdateWidget(RootRestorationScope oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    _loadRootBucketIfNecessary();
+  }
+
+  bool get _needsRootBucketInserted => _ancestorBucket == null;
+
+  bool get _isWaitingForRootBucket {
+    return widget.restorationId != null && _needsRootBucketInserted && !_rootBucketValid;
+  }
+
+  bool _isLoadingRootBucket = false;
+
+  void _loadRootBucketIfNecessary() {
+    if (_isWaitingForRootBucket && !_isLoadingRootBucket) {
+      _isLoadingRootBucket = true;
+      RendererBinding.instance.deferFirstFrame();
+      ServicesBinding.instance.restorationManager.rootBucket.then((RestorationBucket bucket) {
+        _isLoadingRootBucket = false;
+        if (mounted) {
+          ServicesBinding.instance.restorationManager.addListener(_replaceRootBucket);
+          setState(() {
+            _rootBucket = bucket;
+            _rootBucketValid = true;
+            _okToRenderBlankContainer = false;
+          });
+        }
+        RendererBinding.instance.allowFirstFrame();
+      });
+    }
+  }
+
+  void _replaceRootBucket() {
+    _rootBucketValid = false;
+    _rootBucket = null;
+    ServicesBinding.instance.restorationManager.removeListener(_replaceRootBucket);
+    _loadRootBucketIfNecessary();
+    assert(!_isWaitingForRootBucket); // Ensure that load finished synchronously.
+  }
+
+  @override
+  void dispose() {
+    ServicesBinding.instance.restorationManager.removeListener(_replaceRootBucket);
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    if (_okToRenderBlankContainer && _isWaitingForRootBucket) {
+      return Container();
+    }
+
+    return UnmanagedRestorationScope(
+      bucket: _ancestorBucket ?? _rootBucket,
+      child: RestorationScope(
+        restorationId: widget.restorationId,
+        child: widget.child,
+      ),
+    );
+  }
+}
+
+/// Manages an object of type `T`, whose value a [State] object wants to have
```

Comment:
```suggestion
/// Manages an object of type `T`, whose value is a [State] object that it wants to have
```

Reply:
The value is not a State object. The State object wants that value restored.

Reply:
might be clearer (if strictly less technically correct) just saying "Manages an object of type `T` that a [State] object wants to have restored during state restoration" or some such.